File: windows/LinguaLearnAgent/Network/ApiClient.cs
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Threading.Tasks;
using LinguaLearnAgent.Settings;

namespace LinguaLearnAgent.Network;

public class AnalysisPayload
{
    [JsonPropertyName("schemaVersion")]
    public int SchemaVersion { get; set; } = 1;

    [JsonPropertyName("eventId")]
    public string EventId { get; set; } = Guid.NewGuid().ToString();

    [JsonPropertyName("sourceApp")]
    public string SourceApp { get; set; } = "LinguaLearnAgent";

    [JsonPropertyName("originalText")]
    public string OriginalText { get; set; } = string.Empty;

    [JsonPropertyName("text")]
    public string Text { get; set; } = string.Empty;

    [JsonPropertyName("sentAt")]
    public string SentAt { get; set; } = DateTime.UtcNow.ToString("o");

    [JsonPropertyName("previewOnly")]
    public bool PreviewOnly { get; set; } = false;
}

public class AnalysisErrorItem
{
    [JsonPropertyName("original")]
    public string Original { get; set; } = string.Empty;

    [JsonPropertyName("correction")]
    public string Correction { get; set; } = string.Empty;

    [JsonPropertyName("explanationRu")]
    public string ExplanationRu { get; set; } = string.Empty;

    [JsonPropertyName("topic")]
    public string Topic { get; set; } = string.Empty;

    [JsonPropertyName("confidence")]
    public double Confidence { get; set; }

    [JsonPropertyName("level")]
    public string? Level { get; set; }

    [JsonPropertyName("kind")]
    public string? Kind { get; set; }

    [JsonPropertyName("category")]
    public string? Category { get; set; }
}

public class AnalysisEvidenceItem
{
    [JsonPropertyName("topic")]
    public string Topic { get; set; } = string.Empty;

    [JsonPropertyName("outcome")]
    public string Outcome { get; set; } = string.Empty;

    [JsonPropertyName("confidence")]
    public double Confidence { get; set; }

    [JsonPropertyName("explanationRu")]
    public string ExplanationRu { get; set; } = string.Empty;
}

public class AnalysisResponse
{
    [JsonPropertyName("accepted")]
    public bool Accepted { get; set; }

    [JsonPropertyName("schemaVersion")]
    public int SchemaVersion { get; set; }

    [JsonPropertyName("eventId")]
    public string EventId { get; set; } = string.Empty;

    [JsonPropertyName("sourceApp")]
    public string SourceApp { get; set; } = string.Empty;

    [JsonPropertyName("originalText")]
    public string OriginalText { get; set; } = string.Empty;

    [JsonPropertyName("correctedText")]
    public string CorrectedText { get; set; } = string.Empty;

    [JsonPropertyName("recommendedText")]
    public string RecommendedText { get; set; } = string.Empty;

    [JsonPropertyName("changed")]
    public bool Changed { get; set; }

    [JsonPropertyName("assessment")]
    public string Assessment { get; set; } = "correct";

    [JsonPropertyName("hasClearError")]
    public bool HasClearError { get; set; }

    [JsonPropertyName("summaryRu")]
    public string SummaryRu { get; set; } = string.Empty;

    [JsonPropertyName("errors")]
    public List<AnalysisErrorItem> Errors { get; set; } = new();

    [JsonPropertyName("mechanicalCorrections")]
    public List<AnalysisErrorItem> MechanicalCorrections { get; set; } = new();

    [JsonPropertyName("optionalSuggestions")]
    public List<AnalysisErrorItem> OptionalSuggestions { get; set; } = new();

    [JsonPropertyName("topicEvidence")]
    public List<AnalysisEvidenceItem> TopicEvidence { get; set; } = new();

    [JsonPropertyName("previewOnly")]
    public bool PreviewOnly { get; set; }

    [JsonPropertyName("rejectionReason")]
    public string? RejectionReason { get; set; }
}

public class ApiClient
{
    private static readonly JsonSerializerOptions JsonOptions = new JsonSerializerOptions
    {
        PropertyNameCaseInsensitive = true
    };

    private readonly HttpClient _httpClient;
    private readonly PrivacyConsentManager _settings;
    private string _baseUrl = "http://localhost:3001";

    public ApiClient(PrivacyConsentManager settings, HttpClient? customClient = null)
    {
        _settings = settings;
        _httpClient = customClient ?? new HttpClient();
    }

    public void SetBaseUrl(string baseUrl)
    {
        _baseUrl = baseUrl.TrimEnd('/');
    }

    public async Task<AnalysisResponse?> AnalyzeWritingAsync(AnalysisPayload payload)
    {
        try
        {
            string url = $"{_baseUrl}/api/writing/analyze";
            string json = JsonSerializer.Serialize(payload, JsonOptions);
            using var request = new HttpRequestMessage(HttpMethod.Post, url);
            request.Content = new StringContent(json, Encoding.UTF8, "application/json");

            string token = _settings.DeviceToken;
            if (!string.IsNullOrWhiteSpace(token))
            {
                request.Headers.Add("Authorization", $"Bearer {token}");
            }

            using var response = await _httpClient.SendAsync(request);
            if (!response.IsSuccessStatusCode) return null;

            string respJson = await response.Content.ReadAsStringAsync();
            return JsonSerializer.Deserialize<AnalysisResponse>(respJson, JsonOptions);
        }
        catch (Exception ex)
        {
            Console.WriteLine($"[ApiClient] Analysis request failed: {ex.Message}");
            return null;
        }
    }

    public async Task<bool> SendAnalysisAsync(AnalysisPayload payload)
    {
        var result = await AnalyzeWritingAsync(payload);
        return result != null && result.Accepted;
    }
}

File: ios/LinguaLearn/LinguaLearnKeyboardExtension/Network/ApiClient.swift
import Foundation

public struct WritingAnalysisErrorItem: Codable {
    public let kind: String?
    public let category: String?
    public let topic: String?
    public let originalFragment: String?
    public let replacementFragment: String?
    public let message: String?
    public let explanationRu: String?

    enum CodingKeys: String, CodingKey {
        case kind
        case category
        case topic
        case originalFragment = "original_fragment"
        case replacementFragment = "replacement_fragment"
        case message
        case explanationRu = "explanation_ru"
    }

    public init(
        kind: String? = nil,
        category: String? = nil,
        topic: String? = nil,
        originalFragment: String? = nil,
        replacementFragment: String? = nil,
        message: String? = nil,
        explanationRu: String? = nil
    ) {
        self.kind = kind
        self.category = category
        self.topic = topic
        self.originalFragment = originalFragment
        self.replacementFragment = replacementFragment
        self.message = message
        self.explanationRu = explanationRu
    }
}

public struct MechanicalCorrectionItem: Codable {
    public let originalFragment: String?
    public let replacementFragment: String?
    public let explanationRu: String?

    enum CodingKeys: String, CodingKey {
        case originalFragment = "original_fragment"
        case replacementFragment = "replacement_fragment"
        case explanationRu = "explanation_ru"
    }

    public init(originalFragment: String? = nil, replacementFragment: String? = nil, explanationRu: String? = nil) {
        self.originalFragment = originalFragment
        self.replacementFragment = replacementFragment
        self.explanationRu = explanationRu
    }
}

public struct OptionalSuggestionItem: Codable {
    public let suggestion: String?
    public let explanationRu: String?

    enum CodingKeys: String, CodingKey {
        case suggestion
        case explanationRu = "explanation_ru"
    }

    public init(suggestion: String? = nil, explanationRu: String? = nil) {
        self.suggestion = suggestion
        self.explanationRu = explanationRu
    }
}

public struct TopicEvidenceItem: Codable {
    public let topic: String
    public let confidence: Double
    public let outcome: String

    public init(topic: String, confidence: Double, outcome: String) {
        self.topic = topic
        self.confidence = confidence
        self.outcome = outcome
    }
}

public struct AnalysisResponse: Codable {
    public let schemaVersion: Int?
    public let acceptedRaw: Bool?
    public let eventIdRaw: String?
    public let originalTextRaw: String?
    public let correctedText: String
    public let changedRaw: Bool?
    public let summaryRuRaw: String?
    public let previewOnlyRaw: Bool?
    public let rejectionReason: String?
    public let assessment: String?
    public let hasClearError: Bool?
    public let errors: [WritingAnalysisErrorItem]?
    public let mechanicalCorrections: [MechanicalCorrectionItem]?
    public let optionalSuggestions: [OptionalSuggestionItem]?
    public let recommendedText: String?
    public let topicEvidence: [TopicEvidenceItem]?

    public var accepted: Bool { acceptedRaw ?? true }
    public var eventId: String { eventIdRaw ?? "" }
    public var originalText: String { originalTextRaw ?? "" }
    public var changed: Bool { changedRaw ?? (originalText != correctedText) }
    public var summaryRu: String { summaryRuRaw ?? "" }
    public var previewOnly: Bool { previewOnlyRaw ?? false }

    enum CodingKeys: String, CodingKey {
        case schemaVersion
        case acceptedRaw = "accepted"
        case eventIdRaw = "eventId"
        case originalTextRaw = "originalText"
        case correctedText
        case changedRaw = "changed"
        case summaryRuRaw = "summaryRu"
        case previewOnlyRaw = "previewOnly"
        case rejectionReason
        case assessment
        case hasClearError
        case errors
        case mechanicalCorrections
        case optionalSuggestions
        case recommendedText
        case topicEvidence
    }

    public init(
        schemaVersion: Int? = 1,
        accepted: Bool = true,
        eventId: String = "",
        originalText: String = "",
        correctedText: String,
        changed: Bool = true,
        summaryRu: String = "",
        previewOnly: Bool = false,
        rejectionReason: String? = nil,
        assessment: String? = nil,
        hasClearError: Bool? = nil,
        errors: [WritingAnalysisErrorItem]? = nil,
        mechanicalCorrections: [MechanicalCorrectionItem]? = nil,
        optionalSuggestions: [OptionalSuggestionItem]? = nil,
        recommendedText: String? = nil,
        topicEvidence: [TopicEvidenceItem]? = nil
    ) {
        self.schemaVersion = schemaVersion
        self.acceptedRaw = accepted
        self.eventIdRaw = eventId
        self.originalTextRaw = originalText
        self.correctedText = correctedText
        self.changedRaw = changed
        self.summaryRuRaw = summaryRu
        self.previewOnlyRaw = previewOnly
        self.rejectionReason = rejectionReason
        self.assessment = assessment
        self.hasClearError = hasClearError
        self.errors = errors
        self.mechanicalCorrections = mechanicalCorrections
        self.optionalSuggestions = optionalSuggestions
        self.recommendedText = recommendedText
        self.topicEvidence = topicEvidence
    }
}

public class ApiClient {
    public let customBaseUrl: String?
    private let session: URLSession

    public var baseUrl: String {
        return customBaseUrl ?? AppConfig.baseUrl
    }

    public init(baseUrl: String? = nil, session: URLSession? = nil) {
        self.customBaseUrl = baseUrl
        if let session = session {
            self.session = session
        } else {
            let config = URLSessionConfiguration.default
            config.timeoutIntervalForRequest = 3.0
            config.timeoutIntervalForResource = 5.0
            self.session = URLSession(configuration: config)
        }
    }

    public func analyze(
        payload: QueuedWritingPayload,
        deviceToken: String,
        completion: @escaping (Result<AnalysisResponse, Error>) -> Void
    ) {
        guard let url = URL(string: "\(baseUrl)/api/writing/analyze") else {
            completion(.failure(NSError(domain: "ApiClient", code: 400, userInfo: [NSLocalizedDescriptionKey: "Invalid URL"])))
            return
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.timeoutInterval = 3.0
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(deviceToken)", forHTTPHeaderField: "Authorization")

        let bodyDict: [String: Any] = [
            "schemaVersion": payload.schemaVersion,
            "eventId": payload.eventId,
            "sourceApp": payload.sourceApp,
            "originalText": payload.originalText,
            "text": payload.text,
            "sentAt": payload.sentAt,
            "previewOnly": payload.previewOnly
        ]

        request.httpBody = try? JSONSerialization.data(withJSONObject: bodyDict)

        session.dataTask(with: request) { data, response, error in
            if let error = error {
                completion(.failure(error))
                return
            }

            guard let httpResponse = response as? HTTPURLResponse else {
                completion(.failure(NSError(domain: "ApiClient", code: 500, userInfo: [NSLocalizedDescriptionKey: "No response"])))
                return
            }

            guard (200...299).contains(httpResponse.statusCode), let data = data else {
                completion(.failure(NSError(domain: "ApiClient", code: httpResponse.statusCode, userInfo: [NSLocalizedDescriptionKey: "HTTP \(httpResponse.statusCode)"])))
                return
            }

            do {
                let decoded = try JSONDecoder().decode(AnalysisResponse.self, from: data)
                completion(.success(decoded))
            } catch {
                completion(.failure(error))
            }
        }.resume()
    }
}
